plugins/pipette/src/pipette.cc

```diff
@@ -0,0 +1,223 @@
+    gdk_threads_enter();
+    view->setStatusMessage("Error when requesting the image data.");
+    gdk_threads_leave();
+    enabled = true;
+    return;
+  }
+  auto colors = pipette->getPixelAverages(rect);
+
+  // If selection became null the plugin was switched off so ignore the result
+  enabled = true;
+  if(selection == nullptr)
+  {
+    return;
+  }
+
+  std::stringstream info;
+  info.precision(2);
+  fixed(info);
+
+  info << "Top-left: " << rect.getTopLeft();
+  info << ", Bottom-right: " << rect.getBottomRight();
+  info << ", Height: " << rect.getHeight();
+  info << ", Width: " << rect.getWidth();
+  if(!colors.empty()){
+    info << ", Colors:";
+    for(auto element : colors)
+    {
+      info << ' ' << element.first << ": " << element.second;
+    }
+  }
+
+  gdk_threads_enter();
+  view->setStatusMessage(info.str());
+  gdk_threads_leave();
+}
+
+////////////////////////////////////////////////////////////////////////
+// SelectionListener
+////////////////////////////////////////////////////////////////////////
+
+void PipetteHandler::onSelectionStart(GdkPoint, ViewInterface::Ptr)
+{
+}
+
+void PipetteHandler::onSelectionUpdate(Selection::Ptr s, ViewInterface::Ptr view)
+{
+  UNUSED(view);
+  if(enabled)
+  {
+    selection = s;
+  }
+}
+
+void PipetteHandler::onSelectionEnd(Selection::Ptr s, ViewInterface::Ptr view)
+{
+  if(enabled)
+  {
+    selection = s;
+
+    // Prevent more than one job
+    enabled = false;
+    Sequentially()->schedule(boost::bind(&PipetteHandler::computeValues, shared_from_this<PipetteHandler>(), view), pendingJobs);
+  }
+}
+
+////////////////////////////////////////////////////////////////////////
+// PostRenderer
+////////////////////////////////////////////////////////////////////////
+
+void PipetteHandler::render(ViewInterface::Ptr const& vi, cairo_t* cr, Scroom::Utils::Rectangle<double> presentationArea, int zoom)
+{
+  UNUSED(vi);
+
+  if(selection)
+  {
+    auto start = Scroom::Utils::Point<int>(selection->start) - presentationArea.getTopLeft();
+    auto end = Scroom::Utils::Point<int>(selection->end) - presentationArea.getTopLeft();
+
+    if(zoom>=0)
+    {
+      const int pixelSize=1<<zoom;
+      start *= pixelSize;
+      end *= pixelSize;
+    }
+    else
+    {
+      const int pixelSize=1<<-zoom;
+      start /= pixelSize;
+      end /= pixelSize;
+    }
+
+    cairo_set_line_width(cr, 1);
+    cairo_set_source_rgb(cr, 0, 0, 1); // Blue
+    cairo_move_to(cr, end.x, start.y);
+    cairo_line_to(cr, start.x, start.y);
+    cairo_line_to(cr, start.x, end.y);
+    cairo_line_to(cr, end.x, end.y);
+    cairo_line_to(cr, end.x, start.y);
+    cairo_stroke(cr);
+  }
+}
+
+////////////////////////////////////////////////////////////////////////
+// ToolStateListener
+////////////////////////////////////////////////////////////////////////
+
+void PipetteHandler::onDisable(){
+  selection = nullptr;
+  enabled = false;
```

Comment:
.. So enabled has two functions. It deals with whether the tool is enabled, and it has a (more obscure) role in managing concurrency. Right?

Reply:
The `enable` flag toggles the handling of incoming events, so it might require a rename. But yes we want to ignore incoming events in two cases. When the plugin is disabled and while a background computation is in progress.